### PR TITLE
cc: add torch backend support for Multiple backend

### PR DESCRIPTION
**Pull request by CaRoLZhangxy**
need to test in union environment (tf and pt)
see https://github.com/deepmodeling/deepmd-kit/issues/3119

### CODE REVIEW

File: source/api_cc/src/DeepPotPT.cc [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+  if (inited) {
+    std::cerr << "WARNING: deepmd-kit should not be initialized twice, do "
+                 "nothing at the second call of initializer"
+              << std::endl;
+    return;
+  }
+  std::cout << "load model from: " << model << " to gpu " << gpu_rank
+            << std::endl;
+  gpu_id = gpu_rank;
+  torch::Device device(torch::kCUDA, gpu_rank);
```
Comment (njzjz) [inline source/api_cc/src/DeepPotPT.cc:27]: What happens when the GPU is not available?

File: source/api_cc/src/DeepPotPT.cc [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+                                    std::vector<double>& virial,
+                                    std::vector<double>& atom_energy,
+                                    std::vector<double>& atom_virial,
+                                    const int& nframes,
+                                    const std::vector<double>& coord,
+                                    const std::vector<int>& atype,
+                                    const std::vector<double>& box,
+                                    const std::vector<double>& fparam,
+                                    const std::vector<double>& aparam) {
+  throw;
```
Comment (njzjz) [inline source/api_cc/src/DeepPotPT.cc:241]: Always throw `deepmd::deepmd_exception`, which can be caught by the downstream software (e.g., LAMMPS).

File: source/api_cc/include/DeepPotPT.h [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+    assert(inited);
+    return rcut;
+  };
+  /**
+   * @brief Get the number of types.
+   * @return The number of types.
+   **/
+  int numb_types() const {
+    assert(inited);
+    return ntypes;
```
Comment (njzjz) [inline source/api_cc/include/DeepPotPT.h:198]: These variables (`ntypes`, `ntypes_spin`, `dim_fparam`, `dim_aparam`, `aparam_nall`) seem not to be initiated anywhere, so it's expected to see a segmentation fault.

**Review by njzjz — CHANGES_REQUESTED** [review]

File: source/api_cc/src/commonPT.cc [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  int inum = inlist.inum;
+  ilist.resize(inum);
+  numneigh.resize(inum);
+  memcpy(&ilist[0], inlist.ilist, inum * sizeof(int));
+  int* max_element = std::max_element(inlist.numneigh, inlist.numneigh + inum);
+  max_num_neighbors = *max_element;
+  if (max_num_neighbors < nnei) {
+    max_num_neighbors = nnei;
+  }
+  jlist = (int*)malloc(inum * max_num_neighbors * sizeof(int));
```
Comment (njzjz) [inline source/api_cc/src/commonPT.cc:17]: Doesn't it cause the memory leaking since it is never freed?

File: source/api_cc/src/DeepPotPT.cc (hunk 1 above)
Comment (njzjz) [reply source/api_cc/src/DeepPotPT.cc:31]: We may use `torch::cuda::is_available` to check here.

File: deepmd/pt/model/model/dp_atomic_model.py [hunk 5]
```diff
@@ -100,6 +101,10 @@ def get_sel(self) -> List[int]:
         """Get the neighbor selection."""
         return self.sel
 
+    def get_nnei(self) -> List[int]:
```
Comment (njzjz) [inline deepmd/pt/model/model/dp_atomic_model.py:104]: Perhaps [`torch.jit.export`](https://pytorch.org/docs/stable/jit.html#torch.jit.export) should be used?

File: source/api_cc/src/commonPT.cc [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+                                         int& max_num_neighbors) {
+  int inum = inlist.inum;
+  ilist.resize(inum);
+  numneigh.resize(inum);
+  memcpy(&ilist[0], inlist.ilist, inum * sizeof(int));
+  int* max_element = std::max_element(inlist.numneigh, inlist.numneigh + inum);
+  max_num_neighbors = *max_element;
+  unsigned long nlist_size = (unsigned long)inum * max_num_neighbors;
+  jlist.resize(nlist_size);
+  memset(&jlist[0], -1, nlist_size * sizeof(int));
```
Comment (njzjz) [inline source/api_cc/src/commonPT.cc:15]: It seems wrong: https://stackoverflow.com/a/20951450/9567349

**Review by njzjz — CHANGES_REQUESTED** [review]

**Review by njzjz — APPROVED** [review]
fparam and apraram are not supported yet. But I think it's okay for this PR.

File: source/api_cc/src/DeepPotPT.cc [hunk 7]
```diff
@@ -1,8 +1,398 @@
...
+//     ENERGYTYPE& ener,
+//     std::vector<float>& force,
+//     std::vector<float>& virial,
+//     std::vector<float>& atom_energy,
+//     std::vector<float>& atom_virial,
+//     const std::vector<float>& coord,
+//     const std::vector<int>& atype,
+//     const std::vector<float>& box,
+//     const InputNlist& lmp_list,
+//     const int& ago);
```
Comment (wanghan-iapcm) [inline source/api_cc/src/DeepPotPT.cc:170]:
```suggestion

```

File: source/api_cc/src/DeepPotPT.cc [hunk 8]
```diff
@@ -1,8 +1,398 @@
...
+//     const std::vector<double>& box);
+// template void DeepPotPT::compute<float, ENERGYTYPE>(
+//     ENERGYTYPE& ener,
+//     std::vector<float>& force,
+//     std::vector<float>& virial,
+//     std::vector<float>& atom_energy,
+//     std::vector<float>& atom_virial,
+//     const std::vector<float>& coord,
+//     const std::vector<int>& atype,
+//     const std::vector<float>& box);
```
Comment (wanghan-iapcm) [inline source/api_cc/src/DeepPotPT.cc:287]:
```suggestion

```

**Review by wanghan-iapcm — CHANGES_REQUESTED** [review]

**Review by wanghan-iapcm — APPROVED** [review]

**Pull request merged**